sql/src/test/java/io/crate/execution/dml/upsert/SourceFromCellsTest.java

```diff
@@ -250,10 +249,29 @@ public void test_generated_based_on_default() throws Exception {
         DocTableInfo t6 = e.resolveTableInfo("t6");
         InsertSourceFromCells sourceFromCells = new InsertSourceFromCells(
             txnCtx, e.functions(), t6, "t6", GeneratedColumns.Validation.VALUE_MATCH, List.of());
-        BytesReference source = sourceFromCells.generateSource(new Object[0]);
+        BytesReference source = sourceFromCells.generateSourceAndCheckConstraints(new Object[0]);
         Map<String, Object> map = JsonXContent.jsonXContent.createParser(
             NamedXContentRegistry.EMPTY, DeprecationHandler.THROW_UNSUPPORTED_OPERATION, BytesReference.toBytes(source)).map();
         assertThat(Maps.getByPath(map, "x"), is(1));
         assertThat(Maps.getByPath(map, "y"), is(2));
     }
+
+    @Test
+    public void test_apply_not_null_constaint_against_gen_col_value_after_it_was_inserted_into_source() throws Exception {
+        var e = SQLExecutor.builder(clusterService)
+            .addTable("create table t (x int, y as x not null)")
+            .build();
+        DocTableInfo t = e.resolveTableInfo("t");
+        InsertSourceFromCells sourceFromCells = new InsertSourceFromCells(
+            txnCtx,
+            e.functions(),
+            t,
+            "t",
+            GeneratedColumns.Validation.VALUE_MATCH,
+            List.of(Objects.requireNonNull(t.getReference(new ColumnIdent("x")))));
+
+        BytesReference source = sourceFromCells.generateSourceAndCheckConstraints(new Object[]{1});
+
+        assertThat(source.utf8ToString(), is("{\"x\":1,\"y\":1}"));
+    }
```

Comment:
can we add a test case to a target table which is partitioned by a generated column? It looks to me that this case may still not proper handled as the partition value must not be part of the generated source...

Reply:
it is the case on master, the partitioned values if they are produced by generated expressions are part of a source map and it does not break anything. Maybe let's keep this behaviour here and follow up if needed separately?

Reply:
Yes sure.